### DISCUSSION TITLE
memory usage in transient mode

### DISCUSSION THREAD

**Discussion by gofford** [ropensci/targets#602] [Help]
I have a set of 17 targets that each train a model. Memory footprint for the model is constrained to be 32GB in size. The overall footprint of the model + data is 35GB.

The pipeline I've built is regularly exhausting memory on a 128GB machine, despite being set to 'transient' memory mode. I'm not sure why this would be happening so I'm wondering if I'm doing something wrong and/or not accounting for something in the way I've structured some of the targets.

To bypass issues with R-native serialization in LightGBM (https://github.com/microsoft/LightGBM/issues/4208, https://github.com/microsoft/LightGBM/issues/4415, https://github.com/curso-r/treesnip/issues/33) I do not use targets to handle the model objects. Instead, I archive them to disk within a given step and have the fitting step return a file containing the filename. These filenames are passed to subsequent steps in the pipeline, which read the files into memory from disk when needed. This is all handled with bespoke code and _not_ by targets. Functionally, this all works aside from the memory issue despite being somewhat convoluted. It exists like this as a workaround for issues with LightGBM serialization.

My question: in the scenario described above, how is the un-load and re-load handled across steps in `transient` mode? In this mode what, precisely, is transient within a step vs what is persistent in persistent mode? For example, is it only the input and outputs (handled by targets) that are treated as transient? Or do variables and objects from earlier steps linger in memory if they are not explicitly removed?

As an example, if I have a pipeline of the form

```
tar_target(
   name = step1, 
   command = {
       model <- fit_model(data), 
       filename <- write_model_to_disk(model)
       return(filename)
   }
),
tar_target(
   name = step2,
   command = {
      model <- load_model_from_disk(filename)
      prediction <- generate_prediction(data, model)
      return(prediction)
   }
),
tar_target(
   name = step3,
   command = {
      model <- load_model_from_disk(filename)
      diagnostics <- make_diagnostics(data, model)
      return(diagnostics)
   }
)
``` 

is the `model` object treated transiently within each step, such that is removed from memory when the step is completed, or can there be multiple instances of it in memory as it is re-used? 

As I said, everything is working technically aside from using far more memory than expected. Really appreciate the feedback and help getting to this point!

**Comment by wlandau** [comment]
Did you also try `tar_option_set(garbage_collection = TRUE)`? All transient targets get released from their bindings every time a target completes, but garbage collection actually frees up that memory. Garbage collection theoretically is done automatically when the memory is about to be maxed out, but it's worth a shot here.

If each target is only a file target, then I would expect memory consumption due to `targets` alone to be small. Do you encounter the same memory issues with an equivalent analysis without `targets`?

Also, I recommend setting `format = "file"` for `step1` because the output file should be tracked for changes.

**Reply by gofford** [reply]
Yep, `tar_option_set(garbage_collection)` is already set to TRUE. Knowing that the transient targets _should_ get freed is reassuring. Already using `file` for the equivalent of `step1`. 

The memory issue only comes up when running in Targets, and worsens with subsequent targets. Branch 1 uses the expected ~40GB memory, second uses >70GB, and the third uses >90GB, etc.. At that point, the pipeline will exit with an error when it exhausts memory. If I re-start the pipeline it will pick up where it left off, and go through the same steps. It's very odd. I'll reduce it down to a reprex but it's likely a nuance that's associated to my particular problem so hard to reduce it down.

**Reply by wlandau** [reply]
What about large files in a `targets` pipeline that do not use any special lightGBM stuff? Could be that lightGBM is putting weird stuff in the global environment or some other environment that keeps adding up. Could clash with `targets` because `targets` runs in special environments in order to prevent side effects.

**Reply by gofford** [reply]
It could well be something like that. Digging into it further, it looks like targets itself is working as expected. I am seeing similar memory uses both inside and outside of targets when running commands manually. 

It seems like much of the memory inflation is from using `tidymodels` (the `recipe` object alone is 10+GB before fitting). Another alternative is that LightGBM might be leaving C objects in memory.

Can you recommend a way to troubleshoot/visualise which objects are in which environment during a targets run? It would be useful to see what exists inside the special targets env (whatever it might be called!) vs what is ending up in the Global env as a possible avenue to troubleshoot this.

**Reply by wlandau** [reply]
Interactive debugging with `browser()` statements in your functions and/or https://books.ropensci.org/targets/debugging.html#the-debug-option could help. Then you could just run ls() to see the available objects.
